processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -107,35 +177,130 @@ public void close() throws IOException
       metaOut.write(0x1);
       metaOut.write(objectsSorted ? 0x1 : 0x0);
-      metaOut.write(Ints.toByteArray((int) numBytesWritten + 4));
-      metaOut.write(Ints.toByteArray(numWritten));
+      metaOut.write(Ints.toByteArray(Ints.saturatedCast(numBytesWritten + 4)));
+      metaOut.write(Ints.toByteArray(Ints.saturatedCast(numWritten)));
     }
-    finally {
-      metaOut.close();
+  }
+
+  private void closeMultiFiles() throws IOException
+  {
+    headerOutLong.close();
+    Preconditions.checkState(
+        headerOutLong.getCount() == (numWritten * 8),
+        "numWritten[%s] number of rows should have [%s] bytes written to headerOutLong, had[%s]",
+        numWritten,
+        numWritten * 8,
+        headerOutLong.getCount()
+    );
+    Preconditions.checkState(
+        headerOutLong.getCount() < Integer.MAX_VALUE,
+        "Wrote[%s] bytes in header file, which is too many.",
+        headerOutLong.getCount()
+    );
+  }
+
+  /**
+   * Tries to get best value split(number of elements in each value file) which can be expressed as power of 2.
+   *
+   * @return Returns the size of value file splits as power of 2.
+   *
+   * @throws IOException
+   */
+  private int bagSizePower() throws IOException
+  {
+    int avgObjectSize = Ints.saturatedCast((valuesOut.getCount() / numWritten));
+
+    File f = ioPeon.getFile(makeFilename("headerLong"));
+    Preconditions.checkNotNull(f, "header file missing.");
+
+    try (RandomAccessFile headerFile = new RandomAccessFile(f, "r")) {
+      for (int i = 31; i >= 0; --i) {
+        if ((1L << i) * avgObjectSize <= fileSizeLimit) {
+          if (actuallyFits(i, headerFile)) {
+            return i;
+          }
+        }
+      }
+    }
+    throw new ISE(
+        "no value split found with fileSizeLimit [%d], avgObjectSize [%d] while serializing [%s]",
+        fileSizeLimit,
+        avgObjectSize,
+        filenameBase
+    );
+  }
+
+  /**
+   * Checks if candidate value splits can divide value file in such a way no object/element crosses the value splits.
+   *
+   * @param powerTwo   candidate value split expressed as power of 2.
+   * @param headerFile header file.
+   *
+   * @return true if candidate value split can hold all splits.
+   *
+   * @throws IOException
+   */
+  private boolean actuallyFits(int powerTwo, RandomAccessFile headerFile) throws IOException
+  {
+    long lastValueOffset = 0;
+    long currentValueOffset = 0;
+    long valueBytesWritten = valuesOut.getCount();
+    long headerIndex = 0;
+    long bagSize = 1L << powerTwo;
+
+    while (lastValueOffset < valueBytesWritten) {
+
+      if (headerIndex + bagSize <= numWritten) {
+        headerFile.seek((headerIndex + bagSize - 1) * Longs.BYTES);
+        currentValueOffset = headerFile.readLong();
+      } else if (headerIndex < numWritten && numWritten < (headerIndex + bagSize)) {
+        headerFile.seek((numWritten - 1) * Longs.BYTES);
+        currentValueOffset = headerFile.readLong();
+      }
```

Comment:
This could be false if `numWritten < (headerIndex + bagSize)` is false which means above if condition is true (assuming headerIndex is always less than numWritten). I see your point assuming headerIndex< numWritten is not always true and can cause infinite loop. Add an extra check for that.
```
     if (headerIndex >=  numWritten) {
        return true;
      }
```